pkg/controllers/clusters/clusterlifecycle/clusterlifecycle_controller.go

```diff
@@ -0,0 +1,360 @@
+	c.enqueue(mcls)
+}
+
+func (c *Controller) updateCluster(old, cur interface{}) {
+	oldMcls := old.(*clusterapi.ManagedCluster)
+	newMcls := cur.(*clusterapi.ManagedCluster)
+
+	if newMcls.DeletionTimestamp != nil {
+		c.deleteCluster(cur)
+		return
+	}
+
+	// Decide whether discovery has reported a status change.
+	if equality.Semantic.DeepEqual(oldMcls.Status, newMcls.Status) {
+		klog.V(4).Infof("no updates on the status of ManagedCluster %s, skipping syncing", klog.KObj(oldMcls))
+		return
+	}
+
+	klog.V(5).Infof("updating ManagedCluster %q", klog.KObj(oldMcls))
+	c.enqueue(newMcls)
+}
+
+func (c *Controller) deleteCluster(obj interface{}) {
+	mcls, ok := obj.(*clusterapi.ManagedCluster)
+	if !ok {
+		tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			utilruntime.HandleError(fmt.Errorf("couldn't get object from tombstone %#v", obj))
+			return
+		}
+		mcls, ok = tombstone.Obj.(*clusterapi.ManagedCluster)
+		if !ok {
+			utilruntime.HandleError(fmt.Errorf("tombstone contained object that is not a ManagedCluster %#v", obj))
+			return
+		}
+	}
+	klog.V(4).Infof("deleting ManagedCluster %q", klog.KObj(mcls))
+	c.enqueue(mcls)
+}
+
+// runWorker is a long-running function that will continually call the
+// processNextWorkItem function in order to read and process a message on the
+// workqueue.
+func (c *Controller) runWorker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+// processNextWorkItem will read a single work item off the workqueue and
+// attempt to process it, by calling the syncHandler.
+func (c *Controller) processNextWorkItem() bool {
+	obj, shutdown := c.workqueue.Get()
+
+	if shutdown {
+		return false
+	}
+
+	// We wrap this block in a func so we can defer c.workqueue.Done.
+	err := func(obj interface{}) error {
+		// We call Done here so the workqueue knows we have finished
+		// processing this item. We also must remember to call Forget if we
+		// do not want this work item being re-queued. For example, we do
+		// not call Forget if a transient error occurs, instead the item is
+		// put back on the workqueue and attempted again after a back-off
+		// period.
+		defer c.workqueue.Done(obj)
+		var key string
+		var ok bool
+		// We expect strings to come off the workqueue. These are of the
+		// form namespace/name. We do this as the delayed nature of the
+		// workqueue means the items in the informer cache may actually be
+		// more up to date that when the item was initially put onto the
+		// workqueue.
+		if key, ok = obj.(string); !ok {
+			// As the item in the workqueue is actually invalid, we call
+			// Forget here else we'd go into a loop of attempting to
+			// process a work item that is invalid.
+			c.workqueue.Forget(obj)
+			utilruntime.HandleError(fmt.Errorf("expected string in workqueue but got %#v", obj))
+			return nil
+		}
+		// Run the syncHandler, passing it the namespace/name string of the
+		// ManagedCluster resource to be synced.
+		requeueAfter, err := c.syncHandler(key)
```

Comment:
weird logic

Reply:
push back obj to queue with a delay to implement periodically monitoring.

Reply:
Why not using `c.workqueue.AddAfter()` directly? @mars1024 WDYT

Reply:
We could have a check on the return `err`, like `NeedsRequeue(err)`.

Reply:
similar implement in controller-runtime: https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/internal/controller/controller.go#L323